tests/phpunit/includes/SettingsTest.php

```diff
@@ -99,6 +99,32 @@ public function testRegisterChangeListener() {
 		$instance->clearChangeListeners();
 	}
 
+	public function testOverride_ListingToOverride() {
+
+		$changeListener = $this->getMockBuilder( '\SMW\Listener\ChangeListener\ChangeListener' )
+			->disableOriginalConstructor()
+			->getMock();
+
+		$changeListener->expects( $this->once() )
+			->method( 'canTrigger' )
+			->with( $this->equalTo( 'Foo' ) )
+			->will( $this->returnValue( true ) );
+
+		$changeListener->expects( $this->once() )
+			->method( 'setAttrs' )
+			->with( $this->equalTo( [ 'Foo' => 'Foobar' ] ) );
+
+		$changeListener->expects( $this->once() )
+			->method( 'trigger' )
+			->with( $this->equalTo( 'Foo' ) );
```

Comment:
Using the mock API like this binds the settings test to ChangeListener in a way not detectable by static analysis. Doing a rename (using a tool) of one of these methods will cause failure here. Maybe that is fair enough inside of a test for ChangeListener or closely related code, but it is strange for a settings test.

Reply:
>  ChangeListener in a way not detectable by static analysis. 

I'm not worried about any static analysis or tools and for this particular case I'm interested in whether the overridden value is send to a change listener in a form I expect it to do.

Reply:
My comment is not specific to this particular case either.

My concern is that if you write this kind of tests consistently, you end up with a codebase that is both brittle and hard to refactor.

I've seen much worse offenders than this one in SMW. Every time I see a PHPUnit mock returning another PHPUnit mock I die a bit inside. So. Much. Coupling.

Reply:
> Maybe that is fair enough inside of a test for ChangeListener or closely related code, but it is strange for a settings test.

I just had a closer look at this test in my IDE since I did not understand how it worked. Now I see ChangeListener is invoked via `set`. So this is closely related code, making that particular concern not applicable to this test.

Reply:
> Every time I see a PHPUnit mock returning another PHPUnit mock I die a bit inside. So. Much. Coupling.

Yes, I could create a `SpyChangeListener` to only be concerned about the output instead of checking particular methods but then I would have to write another piece of code just to encapsulate the `ChangeListener`. If start doing that I would never finish any code because of encapsulation concerns.